// @shopware-pwa/nuxt3-module/dist/index.mjs
import { defineNuxtModule, addPluginTemplate } from '@nuxt/kit';
import { resolve } from 'path';



// -- Unbuild CommonJS Shims --
import __cjs_url__ from 'url';
import __cjs_path__ from 'path';
import __cjs_mod__ from 'module';
const __filename = __cjs_url__.fileURLToPath(import.meta.url);
const __dirname = __cjs_path__.dirname(__filename);
const require = __cjs_mod__.createRequire(import.meta.url);
const index = defineNuxtModule({
  meta: {
    name: "@shopware/nuxt3",
    configKey: "shopware"
  },
  setup(moduleConfig, nuxt) {
    addPluginTemplate({
      filename: "runtime/shopware.plugin.mjs",
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      src: resolve(__dirname, "../plugin.ts"),
      options: {
        shopwareEndpoint: moduleConfig.shopwareEndpoint ?? "https://demo-frontends.shopware.store",
        shopwareAccessToken: moduleConfig.shopwareAccessToken ?? "SWSCBHFSNTVMAWNZDNFKSHLAYW",
        shopwareApiClient: {
          timeout: moduleConfig.apiClientConfig?.timeout ?? "10000"
        }
      }
    });
    nuxt.options.build.transpile.push("@shopware-pwa/composables-next");
    nuxt.hook("imports:sources", (dirs) => {
      dirs.push({
        from: "@shopware-pwa/composables-next",
        imports: [
          "useAddToCart",
          "useCheckout",
          "useCountries",
          "useCustomerPassword",
          "useCmsElementConfig",
          "useShopwareContext",
          "useSessionContext",
          "useCart",
          "useCartItem",
          "useCategory",
          "useCategorySearch",
          "useNavigation",
          "useNavigationContext",
          "useNavigationSearch",
          "useNotifications",
          "useLandingSearch",
          "useListing",
          "useCms",
          "useCmsBlock",
          "useOrderDetails",
          "useProduct",
          "useProductAssociations",
          "useProductConfigurator",
          "useProductPrice",
          "useProductReviews",
          "useProductSearch",
          "useProductSearchSuggest",
          "useProductWishlist",
          "useSalutations",
          "useUser",
          "useWishlist",
          "usePrice",
          "useCustomerOrders",
          "useAddress",
          "useCmsMeta",
          "useNewsletter",
          "useNavigationContext",
          "useOrderPayment",
          "useBreadcrumbs"
        ]
      });
    });
    nuxt.hook("devtools:customTabs", (iframeTabs) => {
      iframeTabs.push({
        name: "frontends",
        title: "Shopware",
        icon: "fa6-brands:shopware",
        view: {
          type: "iframe",
          src: "https://frontends.shopware.com/"
        }
      });
    });
  }
});

export { index as default };
